Command to update specific indirect dependencies

<!--
  These issues are for **concrete and actionable proposals**. If you just have
  a general problem that you would like to brainstorm, open a Discussion instead (https://github.com/pnpm/pnpm/discussions).
-->

<!--
  If this feature request many people in a company/big team, create a post for your company in the following discussion:
  https://github.com/pnpm/pnpm/discussions/3787
  and link the issue in your post.

  This will help us prioritize issues that affect more people.
-->

- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

**Describe the user story**

Users need to update `caniuse-lite` every month to have only actual polyfills in their JS bundle.

`caniuse-lite` is mostly used as indirect dependency (users have only `autoprefixer` in `dependencies`).

Calling `pnpm up caniuse-lite` causes an error:

``` 
 ERR_PNPM_NO_PACKAGE_IN_DEPENDENCIES  None of the specified packages were found in the dependencies.
```

**Describe the solution you'd like**

It will be nice to have a command or extra option for `pnpm up` to update only `caniuse-lite`.

I think the best option will be to not throw `ERR_PNPM_NO_PACKAGE_IN_DEPENDENCIES ` if we have package in lock file.

**Describe the drawbacks of your solution**

I do not see any for now.

**Describe alternatives you've considered**

Right now Browserslist has [a dirty script](https://github.com/browserslist/browserslist/blob/main/update-db.js#L173-L222) to patch pnpm’s lockfile.

In Yarn 2+ we have `yarn up -R 'caniuse-lite'`, which allowed Browserslist to not have direct script for Yarn 2+.

--- a/.changeset/eight-ads-mate.md
+++ b/.changeset/eight-ads-mate.md
@@ -1,0 +1,5 @@
+---
+"@pnpm/plugin-commands-installation": patch
+---
+
+`pnpm update pkg` should not fail if `pkg` not found as a direct dependency, unless `--depth=0` is passed as a CLI option [#4122](https://github.com/pnpm/pnpm/issues/4122).

--- a/packages/plugin-commands-installation/src/installDeps.ts
+++ b/packages/plugin-commands-installation/src/installDeps.ts
@@ -197,7 +197,7 @@ when running add/update with the --workspace option')
   const updateMatch = opts.update && (params.length > 0) ? createMatcher(params) : null
   if (updateMatch != null) {
     params = matchDependencies(updateMatch, manifest, includeDirect)
-    if (params.length === 0) {
+    if (params.length === 0 && opts.depth === 0) {
       throw new PnpmError('NO_PACKAGE_IN_DEPENDENCIES',
         'None of the specified packages were found in the dependencies.')
     }

--- a/packages/plugin-commands-installation/src/recursive.ts
+++ b/packages/plugin-commands-installation/src/recursive.ts
@@ -275,7 +275,7 @@ export default async function recursive (
         rootDir: opts.workspaceDir,
       } as MutatedProject)
     }
-    if ((mutatedImporters.length === 0) && cmdFullName === 'update') {
+    if ((mutatedImporters.length === 0) && cmdFullName === 'update' && opts.depth === 0) {
       throw new PnpmError('NO_PACKAGE_IN_DEPENDENCIES',
         'None of the specified packages were found in the dependencies of any of the projects.')
     }
@@ -412,7 +412,7 @@ export default async function recursive (
 
   throwOnFail(result)
 
-  if (!result.passes && cmdFullName === 'update') {
+  if (!result.passes && cmdFullName === 'update' && opts.depth === 0) {
     throw new PnpmError('NO_PACKAGE_IN_DEPENDENCIES',
       'None of the specified packages were found in the dependencies of any of the projects.')
   }

--- a/packages/plugin-commands-installation/test/update/recursive.ts
+++ b/packages/plugin-commands-installation/test/update/recursive.ts
@@ -217,6 +217,7 @@ test('recursive update with pattern and name in project', async () => {
     await update.handler({
       ...DEFAULT_OPTS,
       allProjects,
+      depth: 0,
       dir: process.cwd(),
       latest: true,
       lockfileDir,
@@ -388,6 +389,7 @@ test('recursive update in workspace should not add new dependencies', async () =
     await update.handler({
       ...DEFAULT_OPTS,
       ...await readProjects(process.cwd(), []),
+      depth: 0,
       dir: process.cwd(),
       recursive: true,
       workspaceDir: process.cwd(),

--- a/packages/plugin-commands-installation/test/update/recursive.ts
+++ b/packages/plugin-commands-installation/test/update/recursive.ts
@@ -231,6 +231,18 @@ test('recursive update with pattern and name in project', async () => {
   expect(err).toBeTruthy()
   expect(err.code).toBe('ERR_PNPM_NO_PACKAGE_IN_DEPENDENCIES')
 
+  // This should not fail because depth=0 is not specified
+  await update.handler({
+    ...DEFAULT_OPTS,
+    allProjects,
+    dir: process.cwd(),
+    latest: true,
+    lockfileDir,
+    recursive: true,
+    selectedProjectsGraph,
+    workspaceDir: process.cwd(),
+  }, ['this-does-not-exist'])
+
   await update.handler({
     ...DEFAULT_OPTS,
     allProjects,

--- a/packages/plugin-commands-installation/test/update/update.ts
+++ b/packages/plugin-commands-installation/test/update/update.ts
@@ -70,34 +70,44 @@ test('update: fail when both "latest" and "workspace" are true', async () => {
   expect(err.message).toBe('Cannot use --latest with --workspace simultaneously')
 })
 
-test('update: fail when package not in dependencies and depth is 0', async () => {
-  prepare({
-    dependencies: {
-      'peer-a': '1.0.0',
-      'peer-c': '1.0.0',
-    },
+describe('update by package name', () => {
+  beforeAll(async () => {
+    prepare({
+      dependencies: {
+        'peer-a': '1.0.0',
+        'peer-c': '1.0.0',
+      },
+    })
+    await install.handler({
+      ...DEFAULT_OPTS,
+      dir: process.cwd(),
+      workspaceDir: process.cwd(),
+    })
   })
-
-  await install.handler({
-    ...DEFAULT_OPTS,
-    dir: process.cwd(),
-    workspaceDir: process.cwd(),
+  it("should fail when the package isn't in the direct dependencies and depth is 0", async () => {
+    let err!: PnpmError
+    try {
+      await update.handler({
+        ...DEFAULT_OPTS,
+        depth: 0,
+        dir: process.cwd(),
+        sharedWorkspaceLockfile: true,
+        workspaceDir: process.cwd(),
+      }, ['peer-b'])
+    } catch (_err: any) { // eslint-disable-line
+      err = _err
+    }
+    expect(err.code).toBe('ERR_PNPM_NO_PACKAGE_IN_DEPENDENCIES')
+    expect(err.message).toBe('None of the specified packages were found in the dependencies of any of the projects.')
   })
-
-  let err!: PnpmError
-  try {
+  it("shouldn't fail when the package isn't in the direct dependencies", async () => {
     await update.handler({
       ...DEFAULT_OPTS,
-      depth: 0,
       dir: process.cwd(),
       sharedWorkspaceLockfile: true,
       workspaceDir: process.cwd(),
     }, ['peer-b'])
-  } catch (_err: any) { // eslint-disable-line
-    err = _err
-  }
-  expect(err.code).toBe('ERR_PNPM_NO_PACKAGE_IN_DEPENDENCIES')
-  expect(err.message).toBe('None of the specified packages were found in the dependencies of any of the projects.')
+  })
 })
 
 test('update --no-save should not update package.json and pnpm-lock.yaml', async () => {

--- a/packages/plugin-commands-installation/test/update/update.ts
+++ b/packages/plugin-commands-installation/test/update/update.ts
@@ -70,7 +70,7 @@ test('update: fail when both "latest" and "workspace" are true', async () => {
   expect(err.message).toBe('Cannot use --latest with --workspace simultaneously')
 })
 
-test('update: fail when package not in dependencies', async () => {
+test('update: fail when package not in dependencies and depth is 0', async () => {
   prepare({
     dependencies: {
       'peer-a': '1.0.0',
@@ -88,6 +88,7 @@ test('update: fail when package not in dependencies', async () => {
   try {
     await update.handler({
       ...DEFAULT_OPTS,
+      depth: 0,
       dir: process.cwd(),
       sharedWorkspaceLockfile: true,
       workspaceDir: process.cwd(),

--- a/packages/pnpm/test/update.ts
+++ b/packages/pnpm/test/update.ts
@@ -137,20 +137,6 @@ test('recursive update --no-shared-workspace-lockfile', async function () {
   expect(pkg.dependencies?.['foo']).toBe('^100.1.0')
 })
 
-test('update should not install the dependency if it is not present already', async () => {
-  const project = prepare()
-
-  let err!: Error
-  try {
-    await execPnpm(['update', 'is-positive'])
-  } catch (_err: any) { // eslint-disable-line
-    err = _err
-  }
-  expect(err).toBeTruthy()
-
-  await project.hasNot('is-positive')
-})
-
 test('update --latest', async function () {
   const project = prepare()
 

